File: 0090-subsets-ii/0090-subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = [[]]

        def dfs(nums, subset, i):
            if i == len(nums):
                return
            for j in range(i, len(nums)):
                if j>i and nums[j] == nums[j-1]:
                    continue
                subset.append(nums[j])
                res.append(subset.copy())
                dfs(nums, subset, j+1)
                subset.pop()

        dfs(sorted(nums), [], 0)
        return res

File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        /**
                2 2 
               2   ~
            2   ~  2   ~

            1    2    2
          12 12  22
        122   
         
        1    2     3
       12 13 23     
     123     
                                 
         */
        List<List<Integer>> res = new ArrayList<>();
        backtrack(nums, 0, new ArrayList<>(), res);
        return res;
    }

    private void backtrack(int[] nums, int start, List<Integer> curr, List<List<Integer>> res) {
        res.add(new ArrayList<>(curr));
        for (int i = start; i < nums.length; i++) {
            if (i > start && nums[i - 1] == nums[i]) continue;
            curr.add(nums[i]);        
            backtrack(nums, i + 1, curr, res);            
            curr.remove(curr.size() - 1);
        }
    }
}